src/Yi/Braid.hs

```diff
@@ -0,0 +1,578 @@
+  where
+    go !acc []                = acc
+    go acc (Braid t : ts') = go (acc >< t' >< t) ts'
+
+-- | Intersperses the given segment between the 'Braid's. This is
+-- useful when you have a bunch of 'Braid's you just want to separate
+-- with something.
+--
+-- What's more, the result is a single 'Braid'. You can easily
+-- achieve a version that blindly inserts elements to the back by
+-- mapping over the list instead of using this function.
+--
+-- You can think of it as a specialised version of
+-- 'Yi.Braid.intercalate'. Note that what this does __not__ do is
+-- intersperse segments into the underlying chain, you should convert
+-- and use your type's underlying intersperse for that instead.
+intersperse :: (ValidBraid v s a) => s -> [Braid v a] -> Braid v a
+intersperse _ [] = mempty
+intersperse c (t:ts) = go t ts
+  where
+    go !acc [] = acc
+    go acc (t':ts') = go (acc <> (c `cons` t')) ts'
+
+-- | Add a segment in front of a 'Braid'.
+cons :: (ValidBraid v s a) => s -> Braid v a -> Braid v a
+cons c (Braid t) = case viewl t of
+  EmptyL -> Yi.Braid.singleton c
+  Chunk l x :< ts | l < defaultChunkSize -> Braid $ Chunk (l + 1) (c `LL.cons` x) <| ts
+  _ -> Braid $ Chunk 1 (LL.singleton c) <| t
+
+-- | Add a segment in the back of a 'Braid'.
+snoc :: (ValidBraid v s a) => Braid v a -> s -> Braid v a
+snoc (Braid t) c = case viewr t of
+  EmptyR -> Yi.Braid.singleton c
+  ts :> Chunk l x | l < defaultChunkSize -> Braid $ ts |> Chunk (l + 1) (x `LL.snoc` c)
+  _ -> Braid $ t |> Chunk 1 (LL.singleton c)
+
+-- | Turn a single segment into a 'Braid'. 
+-- Consider whether it's worth creating
+-- this, maybe you can use 'cons' or 'snoc' instead?
+singleton :: (ValidBraid v s a) => s -> Braid v a
+singleton c = Braid . T.singleton $ Chunk 1 (LL.singleton c)
+
+-- | @any@ specialised to 'Braid'
+--
+-- Implementation note: this currently just does any by doing 'S.any'
+-- on underlying chunks. We should be able to speed it
+-- up by running it in parallel over multiple chunks.
+any :: (ValidBraid v s a) => (s -> Bool) -> Braid v a -> Bool
+any p = go . fromBraid
+  where
+    go x = case viewl x of
+      EmptyL -> False
+      Chunk _ t :< ts -> LL.any p t || go ts
+
+-- | @all@ specialised to 'Braid'
+--
+-- See the implementation note for 'Yi.Braid.any'.
+all :: (ValidBraid v s a) => (s -> Bool) -> Braid v a -> Bool
+all p = go . fromBraid
+  where
+    go x = case viewl x of
+      EmptyL -> True
+      Chunk _ t :< ts -> LL.all p t && go ts
+
+-- | Filters the segments from the underlying chain
+--
+-- >>> filter (/= 'a') "bac"
+-- "bc"
+filter :: (ValidBraid v s a) => (s -> Bool) -> Braid v a -> Braid v a
+filter p = Braid . go . fromBraid
+  where
+    go t = case viewl t of
+      EmptyL -> T.empty
+      Chunk _ x :< ts -> mkChunk LL.length (LL.filter p x) -| go ts
+
+-- | Maps the segments of the underlying chain.
+map :: (ValidBraid v s a, ValidBraid q t b) => (s -> t) -> Braid v a -> Braid q b
```

Comment:
Also a generalized underlying segment map! @siddhanathan

Reply:
`v /= q` would be an interesting case.